File: src/2025-02-22-00-51-50--a-deepseek-r1-test-220-10-1--2-1-huoshan/go/exercises/practice/tree-building/tree_building.go
package tree

import (
	"fmt"
	"sort"
)

type Record struct {
	ID     int
	Parent int
	// feel free to add fields as you see fit
}

type Node struct {
	ID       int
	Children []*Node
	// feel free to add fields as you see fit
}

func Build(records []Record) (*Node, error) {
	if len(records) == 0 {
		return nil, nil
	}

	// Sort records by ID to ensure we process parents before children
	sort.Slice(records, func(i, j int) bool {
		return records[i].ID < records[j].ID
	})

	nodes := make(map[int]*Node)
	
	for i, record := range records {
		// Validate record ID matches its index after sorting
		if i != record.ID {
			return nil, fmt.Errorf("non-contiguous ID %d", record.ID)
		}

		// Validate root node
		if record.ID == 0 {
			if record.Parent != 0 {
				return nil, fmt.Errorf("root node has parent %d", record.Parent)
			}
			nodes[0] = &Node{ID: 0}
			continue
		}

		// Validate parent exists and has lower ID
		if record.Parent >= record.ID {
			return nil, fmt.Errorf("node %d has invalid parent %d", record.ID, record.Parent)
		}

		parent, exists := nodes[record.Parent]
		if !exists {
			return nil, fmt.Errorf("parent %d not found for node %d", record.Parent, record.ID)
		}

		// Create and append child node
		nodes[record.ID] = &Node{ID: record.ID}
		parent.Children = append(parent.Children, nodes[record.ID])
	}

	return nodes[0], nil
}

File: src/2025-02-22-00-51-50--a-deepseek-r1-test-220-10-1--2-1-huoshan/python/exercises/practice/tree-building/tree_building.py
class Record:
    def __init__(self, record_id, parent_id):
        self.record_id = record_id
        self.parent_id = parent_id


class Node:
    def __init__(self, node_id):
        self.node_id = node_id
        self.children = []


def BuildTree(records):
    if not records:
        return None
        
    records.sort(key=lambda x: x.record_id)
    ordered_id = [i.record_id for i in records]
    
    # Validate ID sequence
    if ordered_id[0] != 0:
        raise ValueError("Record id is invalid or out of order.")
    if ordered_id[-1] != len(records) - 1:
        raise ValueError("Record id is invalid or out of order.")
    if records[-1].record_id != len(records) - 1:
        raise ValueError("Tree IDs must be continuous from 0 to len(records)-1")
    
    nodes = {r.record_id: Node(r.record_id) for r in records}
    root = None
    for record in records:
        node_id = record.record_id
        parent_id = record.parent_id
        
        # Validate root node
        if node_id == 0:
            if parent_id != 0:
                raise ValueError("Node parent_id should be smaller than it's record_id.")
            root = nodes[node_id]
            continue
            
        # Validate parent-child relationships
        if parent_id > node_id:
            raise ValueError("Node parent_id should be smaller than it's record_id.")
        if parent_id == node_id:
            if node_id != 0:
                raise ValueError("Only root should have equal record and parent id.")
        if parent_id not in nodes:
            raise ValueError(f"Node {node_id} references missing parent {parent_id}")
            
        # Add child to parent
        parent_node = nodes[parent_id]
        parent_node.children.append(nodes[node_id])
    
    return root
